addons/web_editor/static/src/js/wysiwyg/widgets/link.js

```diff
@@ -0,0 +1,239 @@
+
+const _t = core._t;
+
+/**
+ * Allows to customize link content and style.
+ */
+const Link = Widget.extend({
+    xmlDependencies: ['/web_editor/static/src/xml/wysiwyg.xml'],
+    events: {
+        'input': '_onAnyChange',
+        'change': '_onAnyChange',
+        'input input[name="url"]': '_onURLInput',
+    },
+
+    /**
+     * @constructor
+     * @param {Boolean} data.isButton - whether if the target is a button element.
+     */
+    init: function (parent, options, editable, data, $button, link) {
+        this.options = options || {};
+        this._super(parent, _.extend({
+            title: _t("Link to"),
+        }, this.options));
+
+        this.data = data || {};
+        this.isButton = this.data.isButton;
+        this.$button = $button;
+
+        this.data.className = this.data.className || "";
+        this.data.iniClassName = this.data.iniClassName || "";
+
+        // Using explicit type 'link' to preserve style when the target is <button class="...btn-link"/>.
+        this.colorsData = [
+            {type: this.isButton ? 'link' : '', label: _t("Link"), btnPreview: 'link'},
+            {type: 'primary', label: _t("Primary"), btnPreview: 'primary'},
+            {type: 'secondary', label: _t("Secondary"), btnPreview: 'secondary'},
+            // Note: by compatibility the dialog should be able to remove old
+            // colors that were suggested like the BS status colors or the
+            // alpha -> epsilon classes. This is currently done by removing
+            // all btn-* classes anyway.
+        ];
+
+        this.editable = editable;
+        this.$editable = $(editable);
+
+        if (link) {
+            const range = document.createRange();
+            range.selectNodeContents(link);
+            this.data.range = range;
+            this.$link = $(link);
+        } else {
+            const selection = editable && editable.ownerDocument.getSelection();
+            this.data.range = selection && selection.rangeCount && selection.getRangeAt(0);
+        }
+
+        if (this.data.range) {
+            this.$link = this.$link || $(OdooEditorLib.getInSelection(this.editable.ownerDocument, 'a'));
+            this.data.iniClassName = this.$link.attr('class') || '';
+            this.colorCombinationClass = false;
+            let $node = this.$link;
+            while ($node.length && !$node.is('body')) {
+                const className = $node.attr('class') || '';
+                const m = className.match(/\b(o_cc\d+)\b/g);
+                if (m) {
+                    this.colorCombinationClass = m[0];
+                    break;
+                }
+                $node = $node.parent();
+            }
+            this.data.className = this.data.iniClassName.replace(/(^|\s+)btn(-[a-z0-9_-]*)?/gi, ' ');
+
+            if (this.$link.length) {
+                const textLink = this.$link[0];
+                const [encodedText, images] = wysiwygUtils.encodeNodeToText(textLink);
```

Comment:
`encodeNodeToText`, `decodeText`, have been moved in `wysiwygUtils` to share the method in LinkDialog and LinkTools.
It make more sense to have the method to this class rather than having them wysiwygUtils.

Reply:
Done.